plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/server/N4JSStatefulIncrementalBuilder.java

```diff
@@ -29,4 +141,173 @@ protected void unloadResource(URI uri) {
+			if (adjModuleSpecifier != null) {
+				return moduleName2Uri.get(adjModuleSpecifier);
+			}
+		}
+		return Collections.emptyList();
+	}
+
+	private List<ImportDeclaration> getImportDeclarations(WorkspaceAwareResourceSet resourceSet, URI uri) {
+		List<ImportDeclaration> result = new ArrayList<>();
+		Resource resource = null;
+		try {
+			resource = resourceSet.getResource(uri, true);
+		} catch (Exception e) {
+			// ignore error during load
+		}
+		if (resource == null) {
+			return result;
+		}
+
+		for (EObject eobj : resource.getContents()) {
+			if (eobj instanceof Script) {
+				Script script = (Script) eobj;
+
+				for (EObject topLevelStmt : script.getScriptElements()) {
+					if (topLevelStmt instanceof ImportDeclaration) {
+						ImportDeclaration impDecl = (ImportDeclaration) topLevelStmt;
+						result.add(impDecl);
+					} else {
+
+						// we know that all import statements are at the beginning of a file
+						break;
+					}
+				}
+			}
+		}
+		return result;
+	}
+
+	private String getAdjustedModuleSpecifierOrNull(String moduleSpecifier, String prjName,
+			Multimap<String, URI> moduleName2Uri) {
+
+		if (moduleName2Uri.containsKey(moduleSpecifier)) {
+			return moduleSpecifier;
+		} else if (moduleSpecifier.startsWith("./")) {
+			moduleSpecifier = moduleSpecifier.substring(2); // remove './'
+		} else if (moduleSpecifier.startsWith(prjName)) {
+			moduleSpecifier = moduleSpecifier.substring(prjName.length() + 1);
+		}
+
+		if (moduleName2Uri.containsKey(moduleSpecifier)) {
+			return moduleSpecifier;
+		}
+
+		URI msUri = URI.createFileURI(moduleSpecifier);
+		String msWithoutExt = URIUtils.trimFileExtension(msUri).toString();
+		if (moduleName2Uri.containsKey(msWithoutExt)) {
+			return msWithoutExt;
+		}
+		return null;
+	}
+
+	/**
+	 * The given list of URIs is sorted by this method to improve the post processing of N4JS started by the
+	 * {@link N4JSPostProcessor}. During post processing, other dependency resources will be processed too in a
+	 * recursive fashion. However, by sorting URIs beforehand here, that will not be necessary since dependencies
+	 * already have been processed. Note however that in case of dependency cycles, sorting cannot avoid the recursive
+	 * post processing of dependencies.
+	 */
+	private List<URI> computeSortedUriClosure(WorkspaceAwareResourceSet resourceSet, String prjName,
+			Multimap<String, URI> moduleName2Uri, Collection<URI> startUris) {
+
+		List<URI> sortedResults = new ArrayList<>(startUris.size());
+		Set<URI> worklist = new LinkedHashSet<>();
+		worklist.addAll(startUris);
+		while (!worklist.isEmpty()) {
+			Iterator<URI> iter = worklist.iterator();
+			URI uri = iter.next();
+			iter.remove();
+			sortedResults.add(uri);
+
+			Collection<URI> importedUris = getImportedUris(resourceSet, prjName, moduleName2Uri, uri);
+			importedUris.removeAll(sortedResults);
+			worklist.addAll(importedUris);
+		}
+
+		return Lists.reverse(sortedResults);
+	}
```

Comment:
Nice that all the above adjustments can now be cleanly implemented side-by-side in `N4JSStatefulIncrementalBuilder`, thanks to your refactoring of the builder infrastructure!

Reply:
Yes, thank you!